Inputs & Outputs, using real DM, etc [GOL-32]

jboolean opened the pull request:
This PR primarily adds very basic inputs and outputs to Nodes and the ability to move them by dragging. 
Consequently, it also modifies the static test data model to be more similar to the latest data model designs. 

It also contains various cleanup, refactoring or tangential features that I did along the way. 
I do plan to update the method of dragging IOs so they don't feel like they're locked on train tracks. This PR may be declined if feedback warrants a lot of changes.

jboolean committed:
Begin to add ios

jboolean committed:
Missing semis

jboolean committed:
setup for supporting io dragging

jboolean committed:
Dragging IO with constraint on rectangle v1

Want to support dragging directly from one side to another without having to traverse the path as if the IO is on a train track

jboolean committed:
Refactor out position conversion into utility method

jboolean committed:
Make text in graph unselectable

was annoying me

jboolean committed:
scope global mixins

jboolean committed:
Move IO styles into their own file + add global typography vars

jboolean committed:
Clean up call to io drag

jboolean committed:
Enums galore!

jboolean committed:
Woops! Use the enum.

RobertWarrenGilmore commented on examples/public/stylesheets/draw/io.less:
```diff
@@ -0,0 +1,15 @@
+.io {
+  circle {
```
Shouldn't this start with `&.input` for consistency with line 7.

jboolean replied:
It's currently overriding it. I'll do that though.

RobertWarrenGilmore commented on examples/public/stylesheets/main.less:
```diff
@@ -1,69 +1,26 @@
-@import "../vendor/fluidable/fluidable.less";
```
Did we ever conclusively figure out what this was doing?

MorganCabral replied:
Fluidable is a less grid framework that I was using for my parser demo way back. I think getting rid of this will break my parser stuff from a while back, but that is fine because im planing on changing it significantly anyway.

jboolean replied:
I actually don't recall deleting this line so I can put it back if you'd like.

RobertWarrenGilmore replied:
@MorganCabral pls

RobertWarrenGilmore replied:
Let's keep your change. Morgan can add this back in later if it's needed again.

RobertWarrenGilmore commented on golle.sublime-project:
```diff
@@ -2,7 +2,15 @@
+      "file_exclude_patterns":[
+        "*.sublime-workspace"
+     ]
 		}
```
Super weird indentation.

jboolean replied:
You're right, not sure what happened there. Fixing.

RobertWarrenGilmore commented on lib/test-data.json:
```diff
@@ -8,45 +8,21 @@
-          "label": "Delta"
-        }
-      ]
+      "outputs": []
```
Why are we getting rid of the output here? Shouldn't we test both inputs and outputs?

jboolean replied:
Laziness. Readding.

RobertWarrenGilmore commented on lib/test-data.json:
```diff
@@ -8,45 +8,21 @@
-        }
-      ]
+      "inputs": [],
+      "outputs": []
```
Why are we getting rid of these?

jboolean replied:
Laziness. Re-added.

RobertWarrenGilmore commented on lib/viewcomponents/io.jsx:
```diff
@@ -0,0 +1,66 @@
+
+        return (
+           <g
+            className={'io ' + ioClass}
```
There's a react addon that handles className concatenation. If we get any classNames much more complicated than this, I would recommend that we use the addon.

RobertWarrenGilmore replied:
http://facebook.github.io/react/docs/class-name-manipulation.html

jboolean replied:
Thanks for pointing that out. I'll use it here, though it's not quite right for this situation. Definitely something I can see being helpful in the future.

jboolean replied:
Thanks for pointing that out. I'll use it here, though it's not quite right for this situation. Definitely something I can see being helpful in the future.

RobertWarrenGilmore commented on lib/viewcomponents/io.jsx:
```diff
@@ -0,0 +1,66 @@
+- y {Number}
+- size (radius) {Number}
+- labelPosition
+ */
```
There is a way to specify property signatures. http://facebook.github.io/react/docs/reusable-components.html#prop-validation

jboolean replied:
Thanks, that's great! Adding.

jboolean committed:
Fix crazy indentation

jboolean committed:
Restore additional IOs in test data

jboolean committed:
Add proper PropTypes in lieu of comment for documentation

jboolean committed:
Use classSet

jboolean committed:
Reindent

jboolean committed:
Add procfile

jboolean committed:
Move to correct place

jboolean committed:
But input specific styles under input class

jboolean committed:
Merge branch 'develop' into ios

jboolean committed:
Remove file for trying to run on heroku

RobertWarrenGilmore commented on golle.sublime-project:
```diff
@@ -1,8 +1,16 @@
 {
```
This file should not be included in the PR. Make sure that you have the latest .gitignore file from the develop branch.

jboolean replied:
Oh, okay. I thought it was helpful to others.

RobertWarrenGilmore replied:
It's easier if each developer keeps his own project configuration, because some of us use weird settings that do things like saving draft copies of our own work.

jboolean replied:
Fixed.

RobertWarrenGilmore commented on lib/enum/ioLabelPosition.js:
```diff
@@ -0,0 +1,11 @@
+  RIGHT: 'right'
+
+  // Possible future inclusions: perpendicular, perpendicular-flipped, hidden
+  // angled
```
I like this idea.

RobertWarrenGilmore commented on lib/main.jsx:
```diff
@@ -19,6 +19,10 @@ var dummyModel = {
         return TEST_DATA;
+    },
+
+    render: function() {
```
Why did you add this?

jboolean replied:
It's a hack to force a re-render after moving the IOs in lieu of writing a proper data flow.

RobertWarrenGilmore commented on lib/test-data.json:
```diff
@@ -8,17 +8,23 @@
-          "label": "Alpha"
+          "label": "Alpha",
+          "percentage": 25,
+          "direction": "W"
```
Shouldn't this use your enums?

jboolean replied:
Impossible. This is a json file, it cannot access javascript variables.

RobertWarrenGilmore commented on lib/test-data.json:
```diff
@@ -57,7 +71,9 @@
         {
           "id": "2-output-g",
-          "label": "My output"
+          "label": "My output",
```
A small, and probably insignificant note - there is currently no requirement or plan to make labels distinct from ids.

jboolean replied:
Oh ok.

jboolean replied:
Labels being removed.

RobertWarrenGilmore commented on lib/utilities/positionUtils.js:
```diff
@@ -0,0 +1,48 @@
+
+    switch(positionData.direction) {
+      case CardinalDirection.NORTH:
+      return {x: amount, y: 0};
```
No line breaks?

jboolean replied:
Ok. Seemed succinct to me.

RobertWarrenGilmore commented on lib/viewcomponents/graph.jsx:
```diff
@@ -27,6 +27,8 @@ var Graph = React.createClass({
                     x={nodeData.x}
                     y={nodeData.y}
+                    inputs={nodeData.inputs}
+                    outputs={nodeData.outputs}
```
Shouldn't these be getters, in light of the new data models?

jboolean replied:
Yes, a lot of things should change in light of the new data models.

jboolean committed:
PR feedback

RobertWarrenGilmore commented on lib/viewcomponents/io.jsx:
```diff
@@ -0,0 +1,78 @@
+
+    labelPosition: React.PropTypes.oneOf(Object.keys(LabelPosition).map(function(k) {
+      return LabelPosition[k];
+    }))
```
I don't really understand what you're doing here. I get the oneOf and the map function, but what's up with Object.keys? I've never seen that.

jboolean replied:
I wish there was a object.values (maybe I oughta make one). I'm trying to
see that it's a value in the enum.

On Sat, Feb 14, 2015, 12:20 AM Robert Gilmore <notifications@github.com>
wrote:

> In lib/viewcomponents/io.jsx
> <https://github.com/twosigma/goll-e/pull/5#discussion_r24712399>:
>
> > +
> > +var IO = React.createClass({
> > +  propTypes: {
> > +    isInput: React.PropTypes.bool.isRequired,
> > +
> > +    label: React.PropTypes.string,
> > +
> > +    x: React.PropTypes.number.isRequired,
> > +
> > +    y: React.PropTypes.number.isRequired,
> > +
> > +    size: React.PropTypes.number,
> > +
> > +    labelPosition: React.PropTypes.oneOf(Object.keys(LabelPosition).map(function(k) {
> > +      return LabelPosition[k];
> > +    }))
>
> I don't really understand what you're doing here. I get the oneOf and the
> map function, but what's up with Object.keys? I've never seen that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twosigma/goll-e/pull/5/files#r24712399>.
>

RobertWarrenGilmore replied:
I still don't understand.

jboolean replied:
See new code.

RobertWarrenGilmore commented on lib/viewcomponents/io.jsx:
```diff
@@ -0,0 +1,78 @@
+  _getLabelPositioningData: function(labelPosition, extraMargin) {
+    var offset = {x: 0, y: 0};
+
+    var totalMargin = 4 + extraMargin;
```
Is this 4 supposed to be ioRadius?

jboolean replied:
Adding constant.

RobertWarrenGilmore commented on lib/viewcomponents/io.jsx:
```diff
@@ -0,0 +1,78 @@
+var React = require('react');
+var mouseDownDrag = require('../utilities/mouseDownDrag');
```
You declare this and then never use it.

jboolean replied:
√

RobertWarrenGilmore commented on lib/viewcomponents/io.jsx:
```diff
@@ -0,0 +1,78 @@
+  },
+
+  _getLabelPositioningData: function(labelPosition, extraMargin) {
+    var offset = {x: 0, y: 0};
```
You declare this and then never use it.

jboolean replied:
√

RobertWarrenGilmore commented on lib/viewcomponents/node.jsx:
```diff
@@ -1,23 +1,68 @@
+
+        return (
+          <IO id={id} isInput={isInput} model={io} 
+          x={position.x} y={position.y} label={io.label} labelPosition={position.labelPosition}
```
Weird indentation.

jboolean replied:
Oh. looks line I didn't fix some of the 4-space indents that were here. Also sublime does not understand how to indent jsx.

RobertWarrenGilmore replied:
You can get a sublime plugin to support jsx.

jboolean replied:
I have. It indents flat though.

On Sat, Feb 14, 2015, 3:05 PM Robert Gilmore <notifications@github.com>
wrote:

> In lib/viewcomponents/node.jsx
> <https://github.com/twosigma/goll-e/pull/5#discussion_r24716871>:
>
> >              </g>
> >          );
> >      },
> >
> > +    _getRenderedIOs: function(ios, isInput) {
> > +      return ios.map(function(io, id) {
> > +        var position = this._getIOPosition(io);
> > +
> > +        var onIOMove = function(event) {
> > +          this._onIOMove(event, io);
> > +        }.bind(this);
> > +
> > +        return (
> > +          <IO id={id} isInput={isInput} model={io}
> > +          x={position.x} y={position.y} label={io.label} labelPosition={position.labelPosition}
>
> You can get a sublime plugin to support jsx.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twosigma/goll-e/pull/5/files#r24716871>.
>

RobertWarrenGilmore commented on lib/viewcomponents/node.jsx:
```diff
@@ -1,23 +1,68 @@
         );
     },
 
+    _getRenderedIOs: function(ios, isInput) {
```
It took me a while to figure out that these are models. It might make it more clear if the naming of these models were changed throughout this file to have names like ioModels, ioModel, nodeModel. This might be something to do later.

jboolean replied:
ok. Fixed.

RobertWarrenGilmore commented on test/data_model/test-dataModel.js:
```diff
@@ -2,13 +2,15 @@
     CardinalPortPosition = require('../../lib/model/cardinalPortPosition.js')
-	port1 = new Port(1,"out",new CardinalPortPosition("N", 50)),
-	port2 = new Port(2,"in", new CardinalPortPosition("S",60)),
+	port1 = new Port(1,"out",new CardinalPortPosition(CardinalDirection.WEST, 50)),
```
spacing

jboolean replied:
This file is out of the scope of this PR.

jboolean committed:
Merge branch 'develop' of https://github.com/twosigma/goll-e into ios

jboolean commented:
Noticed that the data model has no label property. Is that not a thing?

jboolean committed:
Use new data model for rendering

Requires a JSON to DM parser
Removes user-editable json
Displays ids instead of labels, as labels are not part of the DM spec

jboolean committed:
Missing semicolons

jboolean committed:
Update method name

RobertWarrenGilmore commented on .gitignore:
```diff
@@ -102,4 +102,6 @@ jison/*.js
 
 # Ignore Sublime Text project files.
 *.sublime-workspace
+*.sublime-project
```
You don't have to add this. It's already in develop, on a different line.

jboolean replied:
Can't find it.

RobertWarrenGilmore replied:
I derped. You're right.

jboolean committed:
Remove sublime project file

jboolean commented on lib/model/graphParser.js:
```diff
@@ -0,0 +1,89 @@
+};
+
+var GraphParser = function() {
+  this._nodes = {};
```
Actually. IDs are global right? The idea here was to 
- check for duplicate ids
- turn IDs into object references, like in the case of edges.

If IDs are global, there's only need for one id -> object mapping.

RobertWarrenGilmore replied:
IDs are supposed to be scoped (once we have containers), and thus possible to duplicate. But the full address of a node (the combination of its ID and its containers' IDs) should be unique.

RobertWarrenGilmore replied:
So, for instance, there might be a "Container1>Alpha" and a "Container2>Alpha".

jboolean replied:
Ah. Will fix.

jboolean replied:
Is that syntax still correct?

RobertWarrenGilmore replied:
* The angle bracket syntax is still nominally correct, but containers are not yet implemented, so the syntax is not used anywhere yet.
* Since there are not yet any containers, any test data that we have should have unique IDs.
* Global, because we don't have containers yet.

jboolean replied:
This is not just reliant on containers. Think about IOs.

RobertWarrenGilmore replied:
I see. IOs/Ports don't use the `>`. They use `+` and `-`. So there could be "Alpha+myInput" and "Bravo+myInput".

RobertWarrenGilmore replied:
So if you need globally unique IDs, use the full addresses of the elements in question.

jboolean committed:
Drag and drop IOs!!!

jboolean committed:
Clean up big ternaries and fix snapping to places outside the node

jboolean committed:
Move conversion code to PositionUtils

jboolean committed:
I learned something about scope

jboolean committed:
Correct return type doc

jboolean commented on lib/utilities/positionUtils.js:
```diff
@@ -0,0 +1,114 @@
+   * Convert x, y positions between 0 and 1 (fraction of some bound such as a Node) to Cardinal
+   * @method cartesianToCardinal
+   * @param  {CartesianPortPosition} positionData
+   * @return {CartesianPortPosition}
```
Fixed

RobertWarrenGilmore commented on lib/model/node.js:
```diff
@@ -67,14 +67,14 @@ Node.prototype.isPinned = function() {
 /**
  * @method getIns - Getter method for the inputs property.
```
Change the comment to match the actual method name.

jboolean replied:
Oversight. Fixing.

RobertWarrenGilmore commented on lib/model/node.js:
```diff
@@ -67,14 +67,14 @@ Node.prototype.isPinned = function() {
 };
 
 /**
  * @method getOuts - Getter method for the outputs property.
```
method name

jboolean replied:
√

RobertWarrenGilmore commented on lib/model/node.js:
```diff
@@ -50,7 +50,7 @@ Node.prototype.getPos = function() {
  *         position in 2D space.
  */
-Node.prototype.setPos = function(x,y) {
+Node.prototype.setPos = function(x, y) {
```
While you're at it, it would be cool if this were setPos**ition**.

RobertWarrenGilmore replied:
To match the method name in Port, that is.

jboolean replied:
Agreed. Executing…

RobertWarrenGilmore commented on lib/model/node.js:
```diff
@@ -42,7 +42,7 @@ Node.prototype.getID = function() {
   return{
     x: this._x,
-    y: this._y
+    y: this._y
```
What changed?

jboolean replied:
No idea.

RobertWarrenGilmore commented on lib/model/node.js:
```diff
@@ -29,9 +29,9 @@ var Node = function(id, inputs, outputs, styles, metadata) {
+ * @method getId - Getter method for the id property.
  */
-Node.prototype.getID = function() {
+Node.prototype.getId = function() {
```
You could add `getEgo` and `getSuperEgo`, too. :stuck_out_tongue_winking_eye:

RobertWarrenGilmore commented on lib/utilities/positionUtils.js:
```diff
@@ -0,0 +1,114 @@
+var PositionUtils = {};
+
+// Data model scales values up to 0-100.
+var DATA_MODEL_MULTIPLIER = 100.0;
```
I don't think that you need to add the `.0` to the end. JS has only one number type - `int` and `float` are not distinct.

jboolean replied:
True. Changing.

RobertWarrenGilmore commented on lib/viewcomponents/app.jsx:
```diff
@@ -10,9 +10,9 @@ var App = React.createClass({
-                </div>
+                {/*<div id='model_editor'>
+                    //<textarea onChange={this._onModelChange} value={JSON.stringify(this.props.model.getContents(), null, '  ')} />
+                //</div>*/}
```
Why have you commented this out?

jboolean replied:
Because we're now using data model objects not pure JSON objects throughout the system, the model cannot be directly displayed or edited anymore.

RobertWarrenGilmore replied:
Damn. Okay. I hope we get the parser working soon, so that we can add back the editor panel, but with GCL in place of JSON.

RobertWarrenGilmore commented on lib/viewcomponents/graph.jsx:
```diff
@@ -17,18 +17,14 @@ var Graph = React.createClass({
-        var data = this.props.model.getContents();
+        var graph = this.props.model.getGraph();
 
          // For each element of data.nodes, create a node component.
```
The comment doesn't match the variable names.

jboolean replied:
Correcting. Thanks for catching.

RobertWarrenGilmore commented on lib/viewcomponents/graph.jsx:
```diff
@@ -17,18 +17,14 @@ var Graph = React.createClass({
 
     render: function() {
-        var data = this.props.model.getContents();
+        var graph = this.props.model.getGraph();
```
If you replaced the only usage of the old method, then why did you only deprecate it, rather than delete it?

jboolean replied:
I'll delete it.

RobertWarrenGilmore commented on lib/viewcomponents/node.jsx:
```diff
@@ -1,31 +1,109 @@
+      var position = this._getIOPosition(ioModel);
+
+      var onIOMove = function(event) {
+        this._onIOMove(event, ioModel);
```
I don't see this function anywhere.

RobertWarrenGilmore commented on lib/viewcomponents/node.jsx:
```diff
@@ -1,31 +1,109 @@
+    return ioModels.map(function(ioModel, id) {
+      var position = this._getIOPosition(ioModel);
+
+      var onIOMove = function(event) {
```
Where is this used?

jboolean replied:
woops sorry

jboolean committed:
Scope ids, generate global ids and validate them within their scope.

jboolean committed:
Misc PR feedback

RobertWarrenGilmore commented:
I am okay with this PR. I give my vote to go ahead with the merge.
